### PR TITLE
8254935: Deprecate the PSSParameterSpec(int) constructor

### PR DESCRIPTION
**Pull request by valeriepeng**
Can someone help review this update to the PSSParameterSpec class regarding the constructor with int argument and the DEFAULT static field? Just added @Deprecate javadoc tag and caution about their usage as suggested in the bug record.

A CSR will be filed once the wording changes are reviewed.

Thanks,
Valerie

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8254935](https://bugs.openjdk.java.net/browse/JDK-8254935): Deprecate the PSSParameterSpec(int) constructor
 * [JDK-8283650](https://bugs.openjdk.java.net/browse/JDK-8283650): Deprecate the PSSParameterSpec(int) constructor and DEFAULT static field (**CSR**)

### Reviewers
 * [Sean Mullan](https://openjdk.java.net/census#mullan) (@seanjmullan - **Reviewer**) ⚠️ Review applies to 26e31a640cdf69dcd1fe538723f33a0e26886fef

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7913/head:pull/7913` \
`$ git checkout pull/7913`

Update a local copy of the PR: \
`$ git checkout pull/7913` \
`$ git pull https://git.openjdk.java.net/jdk pull/7913/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7913`

View PR using the GUI difftool: \
`$ git pr show -t 7913`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7913.diff">https://git.openjdk.java.net/jdk/pull/7913.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 1]
```diff
@@ -163,7 +174,14 @@ public PSSParameterSpec(String mdName, String mgfName,
      *         PSS encoding
      * @throws    IllegalArgumentException if {@code saltLen} is
      *         less than 0
+     * @deprecated This constructor uses the default values as defined in
+     *         ASN.1 encoding in PKCS#1 except for the salt length. These
```
Comment (wangweij) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:178]: Add "the" before "ASN.1 encoding".

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 2]
```diff
@@ -102,9 +107,15 @@
 
     /**
      * The PSS parameter set with all default values
+     * @deprecated The default values defined in the ASN.1 encoding in
+     *         PKCS#1 may become obsolete as time progresses. Thus, it is
+     *         recommended to explicitly specify all desired parameter
+     *         values with
+     *         {@link #PSSParameterSpec(String, String, AlgorithmParameterSpec, int, int) PSSParameterSpec}.
```
Comment (wangweij) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:114]: We are deprecating a field so I would say "This field uses default values defined in ... which may become...".

Do we need to write "PKCS #1" with a blank inside? Same below.

Also, the "all desired parameter values" phrase is perfect for the constructor below but this is for a field not a method so "parameters" does not make sense to me. How about something like "Instead of using this field, user should create a new `PSSParameterSpec` object by calling..." or we can just not mention it. User would need to create one anyway.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 3]
```diff
@@ -102,9 +107,15 @@
...
      *
      * @since 1.5
      */
+    @Deprecated(since="19", forRemoval=true)
```
Comment (wangweij) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:118]: Do we really want to remove it in the future? It's awkward but probably not so harmful. Removing it may unnecessarily break existing codes. Same with the constructor below.

File: src/java.base/share/classes/sun/security/rsa/PSSParameters.java [hunk 4]
```diff
@@ -81,12 +80,12 @@ protected void engineInit(AlgorithmParameterSpec paramSpec)
 
     @Override
     protected void engineInit(byte[] encoded) throws IOException {
-        // first initialize with the DEFAULT values before
-        // retrieving from the encoding bytes
-        String mdName = DEFAULT.getDigestAlgorithm();
-        MGF1ParameterSpec mgfSpec = (MGF1ParameterSpec) DEFAULT.getMGFParameters();
-        int saltLength = DEFAULT.getSaltLength();
-        int trailerField = DEFAULT.getTrailerField();
+        // first initialize with the ASN.1 DEFAULT defined in PKCS#1 v2.2
```
Comment (wangweij) [inline src/java.base/share/classes/sun/security/rsa/PSSParameters.java:83]: "DEFAULT values"? Also, "PKCS #1".

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 1 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:178]: Sure.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 2 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:114]: Well, I did a quick search, it seems both PKCS#X and PKCS #X are used interchangeably. I use PKCS#X to be consistent within this file. Maybe it does not matter much? I thought we tend to return without the space, e.g. PKCS8EncodedKeySpec.getFormat(). Other wording suggestions incorporated. Thanks~

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 3 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:118]: Hmm, I am under the impression that we tend to use forRemoval=true when deprecating stuff, so users may pay more attention to it. This does not mean that we MUST remove it? Also, marking it for removal would give users more time in case something came up and removal become necessary?

File: src/java.base/share/classes/sun/security/rsa/PSSParameters.java (hunk 4 above)
Comment (valeriepeng) [reply src/java.base/share/classes/sun/security/rsa/PSSParameters.java:83]: Sure.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 2 above)
Comment (wangweij) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:114]: My understanding is that if there is a "#" then there should be a " " before it. However, if the format without the space is already used elsewhere, we can keep using it.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 3 above)
Comment (wangweij) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:118]: You are right that there is no rule that an API deprecated for removal must be removed. OK, you decide what the best way is.

Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java]: It's true that the chance of removing it is not that high (at least based on current situation). So, I removed the forRemoval=true part as you suggested.

**Comment by valeriepeng** [comment]
CSR filed at https://bugs.openjdk.java.net/browse/JDK-8283650
Please review, thanks!

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 5]
```diff
@@ -107,15 +107,15 @@
 
     /**
      * The PSS parameter set with all default values
-     * @deprecated The default values defined in the ASN.1 encoding in
-     *         PKCS#1 may become obsolete as time progresses. Thus, it is
-     *         recommended to explicitly specify all desired parameter
-     *         values with
+     * @deprecated This field uses the default values defined in the ASN.1
+     *         encoding in PKCS#1 which may become obsolete as time progresses.
+     *         Thus, it is recommended to create a new PSSParameterSpec object
```
Comment (wangweij) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:112]: Should `PSSParameterSpec` be inside a `{@code ...}`?

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 6]
```diff
@@ -71,6 +71,11 @@
  *     SaltLength   -- 20
  *     TrailerField -- 1
  *
+ * <p>Its values are based on the default values in the ASN.1 encoding
```
Comment (wangweij) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:74]: Maybe "defined in the ASN.1..."?

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 7]
```diff
@@ -71,6 +71,11 @@
...
+ * from PKCS#1 standard and may become obsolete as time progresses.
```
Comment (wangweij) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:75]: Either "the PKCS#1 standard" or simply "PKCS#1" sounds good, but "PKCS#1 standard" without "the"?

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 8]
```diff
@@ -71,6 +71,11 @@
...
+ * Please do not rely on PSSParameterSpec.DEFAULT unless these values are
```
Comment (wangweij) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:76]: Put `PSSParameterSpec.DEFAULT` in `{@code...}`?

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 5 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:112]: Yes, good catch. Will fix. Thanks!

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 6 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:74]: Sure.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 7 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:75]: Ok, I will add "the".

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 8 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:76]: Yes, will do.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 9]
```diff
@@ -64,13 +64,18 @@
...
  *     mask generation function (mgf) -- "MGF1"
  *     parameters for mgf -- MGF1ParameterSpec.SHA1
  *     SaltLength   -- 20
  *     TrailerField -- 1
  *
+ * <p>Its values are based on the default values defined in the ASN.1 encoding
+ * from the PKCS#1 standard and may become obsolete as time progresses.
+ * Please do not rely on {@code PSSParameterSpec.DEFAULT} unless these
+ * values are really what you want to use.
+ *
```
Comment (wangweij) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:78]: I just read the whole class spec. Since the default values are already included above on lines 38-41, I think there is no need to use so much words and we can just say "defined above".

Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:78]: I am not sure if everyone reads the complete spec from class to fields and methods. Which part do you want to remove and just use "defined above"? Personally, I find "above" kind of vague... It's better to keep the comment near the field. Perhaps we should trim the class spec instead?

Comment (wangweij) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:78]: Almost the whole class spec is copied from PKCS #1 but at the end you have "defined in the ASN.1 encoding from the PKCS#1 standard", which might make people wonder if it is something else. If you don't want to remove anything, how about "defined in the ASN.1 encoding from the PKCS#1 standard above". It's verbose but not confusing.

BTW, on https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/security/spec/PSSParameterSpec.html, the whole "Note" is shown on one line. I guess that's not intended.

Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:78]: Ah, Ok, I think I see what you are saying now. I agree about trimming the class spec. Are you ok with removing line 67-77 (proposed/new version)? This would reduce the duplication of wordings.

Comment (wangweij) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:78]: Good suggestion.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 10]
```diff
@@ -64,13 +64,18 @@
  *   ...  -- Allows for future expansion --
  * }
  * </pre>
- * <p>Note: the PSSParameterSpec.DEFAULT uses the following:
+ * <p>Note: the {@code PSSParameterSpec.DEFAULT} uses the following:
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:67]: Suggest rewording as:
`Note: {@code PSSParameterSpec.DEFAULT} uses the following values:`

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 11]
```diff
@@ -32,7 +32,7 @@
  * as defined in the
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:32]: On line 31, add "the" before "RSASSA-PSS".

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 12]
```diff
@@ -64,13 +64,18 @@
...
  *     message digest  -- "SHA-1"
  *     mask generation function (mgf) -- "MGF1"
  *     parameters for mgf -- MGF1ParameterSpec.SHA1
  *     SaltLength   -- 20
  *     TrailerField -- 1
  *
+ * <p>Its values are based on the default values defined in the ASN.1 encoding
+ * from the PKCS#1 standard and may become obsolete as time progresses.
+ * Please do not rely on {@code PSSParameterSpec.DEFAULT} unless these
+ * values are really what you want to use.
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:77]: Some suggested rewordings: 

`Its values are based on the default values defined in the PKCS#1 standard, as defined above. Some of these defaults are no longer recommended due to advances in cryptanalysis -- see RFC 8017 for more details. Thus, {@code PSSParameterSpec.DEFAULT} is deprecated and should be used at your own risk.`

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 13]
```diff
@@ -102,9 +107,15 @@
 
     /**
      * The PSS parameter set with all default values
+     * @deprecated This field uses the default values defined in the ASN.1
+     *         encoding in PKCS#1 which may become obsolete as time progresses.
+     *         Thus, it is recommended to create a new {@code PSSParameterSpec}
+     *         object with the desired parameter values using
+     *         {@link #PSSParameterSpec(String, String, AlgorithmParameterSpec, int, int) PSSParameterSpec}.
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:114]: Similar rewording suggestion here: 

"This field uses the default values defined in the PKCS#1 standard. Some of these defaults are no longer recommended due to advances in cryptanalysis -- see RFC 8017 for more details. Thus ..."

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 14]
```diff
@@ -163,7 +174,14 @@ public PSSParameterSpec(String mdName, String mgfName,
      *         PSS encoding
      * @throws    IllegalArgumentException if {@code saltLen} is
      *         less than 0
+     * @deprecated This constructor uses the default values defined in
+     *         the ASN.1 encoding in PKCS#1 except for the salt length.
+     *         These default values may become obsolete as time progresses.
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:179]: Similar rewording suggestion as above.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 10 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java]: This part has been removed to reduce duplication since there is already javadoc for DEFAULT below.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 11 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:32]: Ok.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 12 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java]: Thanks for the rewording suggestion. This part has been removed.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 14 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:168]: Sure.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 13 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java]: Hmm, RFC 8017 is PKCS#1 v2.2 and I assume you mean PKCS#1 v2.2 with link to RFC 8017.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 15]
```diff
@@ -94,17 +88,26 @@
...
     /**
-     * The {@code TrailerFieldBC} constant as defined in PKCS#1
+     * The {@code TrailerFieldBC} constant as defined in the PKCS #1 standard.
      *
      * @since 11
      */
     public static final int TRAILER_FIELD_BC = 1;
 
     /**
      * The PSS parameter set with all default values
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:98]: Nit - add period at end of sentence.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 16]
```diff
@@ -94,17 +88,26 @@
...
+     * @deprecated This field uses the default values defined in the PKCS #1
+     *         standard. Some of these defaults are no longer recommended due
+     *         to advances in cryptanalysis -- see the
+     *         <a href="https://tools.ietf.org/rfc/rfc8017.txt">PKCS#1 v2.2</a>
+     *         standard for more details. Thus, it is recommended to create
+     *         a new {@code PSSParameterSpec} with the desired parameter values
+     *         using
+     *         {@link #PSSParameterSpec(String, String, AlgorithmParameterSpec, int, int) PSSParameterSpec}.
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:106]: I think it would be more clear to see the full signature of the ctor that you are recommending be used instead, so I would change these 2 lines to:

`using the {@link #PSSParameterSpec(String, String, AlgorithmParameterSpec, int, int)} constructor.`

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 17]
```diff
@@ -157,13 +160,21 @@ public PSSParameterSpec(String mdName, String mgfName,
...
      * @throws    IllegalArgumentException if {@code saltLen} is
      *         less than 0
+     * @deprecated This constructor uses the default values defined in
+     *         the PKCS #1 standard except for the salt length. Some of these
+     *         defaults are no longer recommended due to advances in
+     *         cryptanalysis -- see the
+     *         <a href="https://tools.ietf.org/rfc/rfc8017.txt">PKCS#1 v2.2</a>
+     *         standard for more details. Thus, it is recommended to explicitly
+     *         specify all desired parameter values with
+     *         {@link #PSSParameterSpec(String, String, AlgorithmParameterSpec, int, int) PSSParameterSpec}.
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:175]: Same comment about seeing the full signature of the ctor as mentioned above.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 15 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:98]: Sure.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 16 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:106]: Makes sense.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 17 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:175]: Yes.

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 18]
```diff
@@ -95,15 +95,15 @@
...
     /**
-     * The PSS parameter set with all default values
+     * The PSS parameter set with all default values.
      * @deprecated This field uses the default values defined in the PKCS #1
      *         standard. Some of these defaults are no longer recommended due
      *         to advances in cryptanalysis -- see the
      *         <a href="https://tools.ietf.org/rfc/rfc8017.txt">PKCS#1 v2.2</a>
      *         standard for more details. Thus, it is recommended to create
      *         a new {@code PSSParameterSpec} with the desired parameter values
      *         using
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:105]: s/using/using the/

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java [hunk 19]
```diff
@@ -172,7 +172,7 @@ public PSSParameterSpec(String mdName, String mgfName,
      *         <a href="https://tools.ietf.org/rfc/rfc8017.txt">PKCS#1 v2.2</a>
      *         standard for more details. Thus, it is recommended to explicitly
      *         specify all desired parameter values with
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/spec/PSSParameterSpec.java:174]: s/with/with the/

**Review by seanjmullan — APPROVED** [review]

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 18 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java]: Yes, thanks. Need to re-read the whole sentence again even with minor changes...

File: src/java.base/share/classes/java/security/spec/PSSParameterSpec.java (hunk 19 above)
Comment (valeriepeng) [reply src/java.base/share/classes/java/security/spec/PSSParameterSpec.java]: Yes.

**Comment by valeriepeng** [comment]
Thanks Sean and Max for the very thorough review~

**Comment by valeriepeng** [comment]
Thanks Sean and Max for the very thorough review~
